File: SortVision/public/code/quick/ruby/quickSort.rb
# Quick Sort Implementation in Ruby

# 🧠 Swap elements in the array
def swap(arr, i, j)
  arr[i], arr[j] = arr[j], arr[i]
end

# 🔄 Partition the array using Lomuto's scheme (last element as pivot)
def partition(arr, low, high)
  pivot = arr[high]
  i = low - 1

  (low...high).each do |j|
    if arr[j] <= pivot
      i += 1
      swap(arr, i, j)
    end
  end

  swap(arr, i + 1, high)
  i + 1
end

# 🚀 Quick Sort recursive function
def quick_sort(arr, low = 0, high = nil)
  high ||= arr.length - 1
  if low < high
    pi = partition(arr, low, high)
    quick_sort(arr, low, pi - 1)
    quick_sort(arr, pi + 1, high)
  end
  arr
end

# 🧪 Test utility to print array
def print_array(label, arr)
  puts "#{label}: #{arr.join(', ')}"
end

# ▶️ Example usage and test cases
if __FILE__ == $0
  puts "📦 Quick Sort Test Cases"

  # Test Case 1: Normal unsorted array
  arr1 = [10, 7, 8, 9, 1, 5]
  print_array("Original", arr1)
  quick_sort(arr1)
  print_array("Sorted", arr1)
  puts

  # Test Case 2: Already sorted array
  arr2 = [1, 2, 3, 4, 5]
  print_array("Already Sorted", arr2)
  quick_sort(arr2)
  print_array("After Quick Sort", arr2)
  puts

  # Test Case 3: Single element array
  arr3 = [42]
  print_array("Single Element", arr3)
  quick_sort(arr3)
  print_array("After Quick Sort", arr3)
  puts

  # Test Case 4: Empty array
  arr4 = []
  print_array("Empty Array", arr4)
  quick_sort(arr4)
  print_array("After Quick Sort", arr4)
  puts

  # Test Case 5: Array with duplicates
  arr5 = [5, 1, 3, 5, 2, 1, 3]
  print_array("With Duplicates", arr5)
  quick_sort(arr5)
  print_array("After Quick Sort", arr5)
  puts
end

File: SortVision/public/code/quick/c/quickSort.c
#include <stdio.h>

// Function to swap two integers
void swap(int* a, int* b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

// Lomuto partition function: selects the last element as pivot
int partition(int arr[], int low, int high) {
    int pivot = arr[high]; // Pivot element
    int i = low - 1;        // Index of smaller element

    for (int j = low; j < high; j++) {
        // If current element is smaller than or equal to pivot
        if (arr[j] <= pivot) {
            i++;
            swap(&arr[i], &arr[j]);
        }
    }

    // Place pivot at correct position
    swap(&arr[i + 1], &arr[high]);
    return i + 1;
}

// Recursive Quick Sort function using divide and conquer
void quickSort(int arr[], int low, int high) {
    if (low < high) {
        // pi is partitioning index
        int pi = partition(arr, low, high);

        // Recursively sort elements before and after partition
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

// Utility function to print an array
void printArray(int arr[], int size) {
    for (int i = 0; i < size; i++)
        printf("%d ", arr[i]);
    printf("\n");
}

// Main function to test the Quick Sort implementation
int main() {
    // Test Case 1: General array
    int arr1[] = {10, 7, 8, 9, 1, 5};
    int n1 = sizeof(arr1) / sizeof(arr1[0]);
    printf("Original array: ");
    printArray(arr1, n1);
    quickSort(arr1, 0, n1 - 1);
    printf("Sorted array: ");
    printArray(arr1, n1);
    printf("\n");

    // Test Case 2: Already sorted array
    int arr2[] = {1, 2, 3, 4, 5};
    int n2 = sizeof(arr2) / sizeof(arr2[0]);
    quickSort(arr2, 0, n2 - 1);
    printf("Sorted already sorted array: ");
    printArray(arr2, n2);
    printf("\n");

    // Test Case 3: Single element
    int arr3[] = {42};
    quickSort(arr3, 0, 0);
    printf("Single element array: ");
    printArray(arr3, 1);
    printf("\n");

    // Test Case 4: Empty array
    int arr4[] = {};
    quickSort(arr4, 0, -1);
    printf("Empty array: ");
    printArray(arr4, 0);
    printf("\n");

    return 0;
}

/*
🧠 Time Complexity:
- Best Case: O(n log n)
- Average Case: O(n log n)
- Worst Case: O(n^2) — when pivot is the smallest/largest every time

💾 Space Complexity:
- O(log n) auxiliary space for recursion stack (in-place sorting)

📌 Pivot Strategy Used:
- Lomuto scheme (last element as pivot)

🚀 Performance Optimization Notes:
- Can use Median-of-Three or Random Pivot to improve worst-case performance.
- Tail call elimination can reduce recursion stack height.

✅ Edge Cases Handled:
- Empty array
- Single-element array
- Already sorted array
- Repeated elements
*/
